pkg/client/client_venafi_cloud.go

```diff
@@ -0,0 +1,375 @@
+func (c *VenafiCloudClient) PostDataReadings(_ string, _ string, readings []*api.DataReading) error {
+	// orgID and clusterID are ignored in Venafi Cloud auth
+
+	payload := api.DataReadingsPost{
+		AgentMetadata:  c.agentMetadata,
+		DataGatherTime: time.Now().UTC(),
+		DataReadings:   readings,
+	}
+	data, err := json.Marshal(payload)
+	if err != nil {
+		return err
+	}
+
+	if !strings.HasSuffix(c.uploadPath, "/") {
+		c.uploadPath = fmt.Sprintf("%s/", c.uploadPath)
+	}
+	res, err := c.Post(filepath.Join(c.uploadPath, c.uploaderID), bytes.NewBuffer(data))
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	if code := res.StatusCode; code < 200 || code >= 300 {
+		errorContent := ""
+		body, err := io.ReadAll(res.Body)
+		if err == nil {
+			errorContent = string(body)
+		}
+		return fmt.Errorf("received response with status code %d. Body: %s", code, errorContent)
+	}
+
+	return nil
+}
+
+// Post performs an HTTP POST request.
+func (c *VenafiCloudClient) Post(path string, body io.Reader) (*http.Response, error) {
+	token, err := c.getValidAccessToken()
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, fullURL(c.baseURL, path), body)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("Content-Type", "application/json")
+
+	if len(token.accessToken) > 0 {
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token.accessToken))
+	}
+
+	return c.client.Do(req)
+}
+
+// getValidAccessToken returns a valid access token. It will fetch a new access
+// token from the auth server in case the current access token does not exist
+// or it is expired.
+func (c *VenafiCloudClient) getValidAccessToken() (*venafiCloudAccessToken, error) {
+	if c.accessToken == nil || time.Now().Add(time.Minute).After(c.accessToken.expirationTime) {
+		err := c.updateAccessToken()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return c.accessToken, nil
+}
+
+func (c *VenafiCloudClient) updateAccessToken() error {
+	jwtToken, err := c.generateAndSignJwtToken()
+	if err != nil {
+		return err
+	}
+
+	values := url.Values{}
+	values.Set("grant_type", requiredGrantType)
+	values.Set("assertion", jwtToken)
+
+	tokenURL := fullURL(c.baseURL, accessTokenEndpoint)
+	if err != nil {
+		return err
+	}
+
+	encoded := values.Encode()
+	request, err := http.NewRequest(http.MethodPost, tokenURL, strings.NewReader(encoded))
+	if err != nil {
+		return err
+	}
+
+	request.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	request.Header.Add("Content-Length", strconv.Itoa(len(encoded)))
+
+	now := time.Now()
+	accessToken := accessTokenInformation{}
+	err = c.sendHTTPRequest(request, &accessToken)
+	if err != nil {
+		return err
+	}
+
+	c.lock.Lock()
+	c.accessToken = &venafiCloudAccessToken{
+		accessToken:    accessToken.AccessToken,
+		expirationTime: now.Add(time.Duration(accessToken.ExpiresIn) * time.Second),
+	}
+	c.lock.Unlock()
```

Comment:
Hey, passing by: why is `c.accessToken` only locked during the write operation but not locked during read operations?

As it is, this mutex does protect against two simultaneous updates (since two HTTP requests might trigger `updateAccessToken` simultaneously, but another HTTP request might be reading `c.accessToken` while it is being written, defeating the purpose of the mutex. Note that we wouldn't be able to catch this problem with `go tests -race` because the client doesn't have unit tests yet.